src/common/config_parser.h

```diff
@@ -16,48 +13,74 @@
+namespace cli {
+enum struct mode { training, translation, scoring };
+}  // namespace cli
 
+/**
+ * @brief Command-line options parser
+ *
+ * New options and aliases should be defined within `addOptions*` methods.
+ */
 class ConfigParser {
 public:
-  ConfigParser(int argc, char** argv, ConfigMode mode, bool validate = false)
-      : mode_(mode),
-        cmdline_options_("Allowed options", guess_terminal_width()) {
+  ConfigParser(int argc, char** argv, cli::mode mode, bool validate = false)
+      : mode_(mode) {
     parseOptions(argc, argv, validate);
   }
 
+  /**
+   * @brief Parse command-line options
+   *
+   * Options are parsed in the following order, later config options overwrite
+   * earlier:
+   *  * predefined default values
+   *  * options from the config files provided with --config, from left to right
+   *  * options from the model config file, e.g. model.npz.yml
+   *  * aliases expanded into options, e.g. --best-deep
+   *  * options provided as command-line arguments
+   *
+   * Parsed options are available from getConfig().
+   *
+   * @param argc
+   * @param argv
+   * @param validate Do or do not validate parsed options
+   */
   void parseOptions(int argc, char** argv, bool validate);
 
   YAML::Node getConfig() const;
   std::vector<DeviceId> getDevices();
 
+  // Check if the config has a boolean option for the given key and it is
+  // enabled
+  bool has(const std::string& key) const;
+
 private:
-  ConfigMode mode_;
-  boost::program_options::options_description cmdline_options_;
+  cli::mode mode_;
   YAML::Node config_;
 
-  bool has(const std::string& key) const;
+  void addOptionsGeneral(cli::CLIWrapper&);
+  void addOptionsModel(cli::CLIWrapper&);
+  void addOptionsTraining(cli::CLIWrapper&);
+  void addOptionsValidation(cli::CLIWrapper&);
+  void addOptionsTranslation(cli::CLIWrapper&);
+  void addOptionsScoring(cli::CLIWrapper&);
 
-  template <typename T>
-  T get(const std::string& key) const {
-    return config_[key].as<T>();
-  }
+  void addSuboptionsDevices(cli::CLIWrapper&);
+  void addSuboptionsBatching(cli::CLIWrapper&);
+  void addSuboptionsInputLength(cli::CLIWrapper&);
+
+  void expandAliases(cli::CLIWrapper&);
 
-  void addOptionsCommon(boost::program_options::options_description&);
-  void addOptionsModel(boost::program_options::options_description&);
-  void addOptionsTraining(boost::program_options::options_description&);
-  void addOptionsRescore(boost::program_options::options_description&);
-  void addOptionsValid(boost::program_options::options_description&);
-  void addOptionsTranslate(boost::program_options::options_description&);
+  // Change relative paths to absolute paths relative to the config file's
+  // directory
+  void makeAbsolutePaths(const std::vector<std::string>&);
 
-  void validateOptions() const;
-  void validateDevices() const;
+  // Create a new config from all provided config files
+  YAML::Node loadConfigFiles(const std::vector<std::string>&);
```

Comment:
I don't understand this comment. What is that new config? How is it created from the config files? Is this just loading multiple config files into a YAML::Node? Maybe change the comment then.

The problem is that tehre are so many config classes in here, plus "configuration" can refer to the high-level concept of the entire program's configuration, and/or any of its representations. Let us avoid the term "config" and always qualify it what kind of config. E.g. "new Options instance"

Reply:
Concerning second paragraph, we will address this in a different phrase. This is a first step.

Reply:
Lets add a @TODO at an appropriate place in teh soruce code, then Resolve this.